### PR TITLE
[PQ-281][PQ-282] Moderator/Referral

### PR DESCRIPTION
**Pull request by YulianH**

### CODE REVIEW

File: tests/api_tests/api_referal_test.spec.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    test('Referral CRUD ',async () => {
+        const apiContext = await request.newContext()
+        const api = new Api(apiContext)
+        await api.referalPage.setReferal(apiUrl.qaEnvUrl, user.userToken, user.id, referralUser.id)
+        await api.referalPage.getReferalUsers(apiUrl.qaEnvUrl, user.userToken, referralUser.id)
+        await api.referalPage.getReferalRulsList(apiUrl.qaEnvUrl, user.userToken)
+        await api.referalPage.getReferralUsersStatistics(apiUrl.qaEnvUrl, user.userToken, "week")
+        
+    })
+
```
Comment (IvanK-QX) [inline tests/api_tests/api_referal_test.spec.ts:31]: @YulianH remove useless rows

File: pages/apiPages/referal_page.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    async getReferalRulsList(url: string, userToken: string) {
+        const apiContext = await request.newContext({ignoreHTTPSErrors: true})
+        const headers = Headers.userHeader(userToken)
+
+        const apiRequest = await apiContext.get(`${url}:3000/referal/rules/list`, {headers: headers})
+        expect(apiRequest.ok()).toBeTruthy()
+        console.log(`The Rules List is displayed`)  
+    }
+
+    async getReferralUsersStatistics (url: string, userToken: string, period: string) {
```
Comment (IvanK-QX) [inline pages/apiPages/referal_page.ts:47]: @YulianH it's better to use predefine values for period
like period: 'day' | 'week' | 'month

File: tests/api_tests/api_moderator_test.spec.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import { request, test } from "@playwright/test";
+import { apiUrl } from "../../utils/apiUrl";
+import { Api } from "../../pages/Api";
+import { apiDataSet } from "../../utils/dataSet";
+
+let guestUser, admin, createdNewAdmin, newAdmin, user, referralUser, pendingAvatarId, newStream, streamAction, currentStatus
```
Comment (IvanK-QX) [inline tests/api_tests/api_moderator_test.spec.ts:6]: @YulianH pendingAvatarId should be a const, you use it only in one test
createdNewAdmin should be a const, used only in precondition
guestUser const > precondition only
pendingAvatarId  const > one test only
streamAction  const > one test only
currentStatus const > one test only

File: tests/api_tests/api_moderator_test.spec.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+test.describe.only('API test ',async () => {
```
Comment (IvanK-QX) [inline tests/api_tests/api_moderator_test.spec.ts:8]: @YulianH remove .only, affects all tests

File: pages/apiPages/moderator_page.ts [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+
+        const apiRequest = await apiContext.post(`${url}:3000/admin/agent/referalEarnings`, {data, headers: headers})
+        expect(apiRequest.ok()).toBeTruthy()
+        const response = await apiRequest.json()
+        const returnedUserId = response.user._id
+        expect(userId).toEqual(returnedUserId)
+        console.log(`Correct ${returnedUserId} is displayed`)
+    }
+
+    async setAdminProfileStatus(url: string, adminToken: string, referralUserId: string, currentStatus: string, newStatus: string) {
```
Comment (IvanK-QX) [inline pages/apiPages/moderator_page.ts:65]: @YulianH newStatus should be a predefine value

File: pages/apiPages/moderator_page.ts [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+        const apiRequest = await apiContext.post(`${url}:3000/agent/profile`, {data, headers: headers})
+        expect(apiRequest.ok()).toBeTruthy()
+        const response = await apiRequest.json()
+        const returnedAgentUserId = response._id
+        expect(returnedAgentUserId).toEqual(userId)
+        console.log(`The Profile with id ${returnedAgentUserId} is displayed`)
+        return returnedAgentUserId
+    }
+
+    async AdminProfileUpdate(url: string, adminToken: string, userId: string, action: string, userName: string, payoneerEmail: string) {
```
[the remark this answers is not in the document]
Comment (IvanK-QX) [reply pages/apiPages/moderator_page.ts:216]: action should be predefined value

File: tests/api_tests/api_moderator_test.spec.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    
+    test('Admin Profile Get/Update',async () => {
+        const apiContext = await request.newContext()
+        const api = new Api(apiContext)
+        currentStatus = await api.moderatorPage.getAdminProfile(apiUrl.qaEnvUrl, newAdmin.newAdminToken, user.id)
+        await api.moderatorPage.setAdminProfileStatus(apiUrl.qaEnvUrl, newAdmin.newAdminToken, referralUser.id, currentStatus.userStatus, "Blocked")
+        await api.moderatorPage.getAdminProfileList(apiUrl.qaEnvUrl, admin.adminToken)
+        await api.moderatorPage.getAgentProfile(apiUrl.qaEnvUrl, newAdmin.newAdminToken, referralUser.id)
+        await api.moderatorPage.AdminProfileUpdate(apiUrl.qaEnvUrl, newAdmin.newAdminToken, referralUser.id, "updateFields", apiDataSet.randomName, apiDataSet.randomEmail)
+        await api.moderatorPage.adminSetPayoutEmail(apiUrl.qaEnvUrl, newAdmin.newAdminToken, referralUser.id, "test@test.com")
```
Comment (IvanK-QX) [inline tests/api_tests/api_moderator_test.spec.ts:46]: @YulianH instead email you should use variable from data set, avoid hardcoded values

File: tests/api_tests/api_moderator_test.spec.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        await api.moderatorPage.adminApproveAvatar(apiUrl.qaEnvUrl, newAdmin.newAdminToken, pendingAvatarId.avatarId)
+        pendingAvatarId = await api.moderatorPage.getAdminAvatarsApprovalQueue(apiUrl.qaEnvUrl, newAdmin.newAdminToken)
+        await api.moderatorPage.adminDeclineAvatar(apiUrl.qaEnvUrl, newAdmin.newAdminToken, pendingAvatarId.avatarId)
+        
+    }) 
+
+    test('Admin Moderator Actions',async () => {
+        const apiContext = await request.newContext()
+        const api = new Api(apiContext)
+        streamAction = await api.moderatorPage.adminModeratorAction(apiUrl.qaEnvUrl, newAdmin.newAdminToken, newStream.myStreamId, "closedCamera/emptyRoom")
```
Comment (IvanK-QX) [inline tests/api_tests/api_moderator_test.spec.ts:64]: @YulianH add predefined value list to adminModeratorAction method

**Review by IvanK-QX — CHANGES_REQUESTED** [review]

File: pages/Api.ts [hunk 9]
```diff
@@ -10,6 +10,10 @@ import { ApiFollowingPage } from "./apiPages/following_page"
 import { ApiPayoutPage } from "./apiPages/payout_page"
 import { ApiClientSettingsPage } from "./apiPages/clientSettings_page"
 import { ApiSalaryRulesPage } from "./apiPages/salaryRules_page"
+import { ApiModeratorPage } from "./apiPages/moderator_page"
+import { ApiReferalPage } from "./apiPages/referal_page"
+
+
```
[the remark this answers is not in the document]
Comment (YulianH) [reply pages/Api.ts:16]: done

File: pages/apiPages/moderator_page.ts [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+                `${referralUserId}`
+            ]
+        }
+        const headers = Headers.userHeader(adminToken)
+
+        const apiRequest = await apiContext.post(`${url}:3000/admin/profile/status`, {data, headers: headers})
+        expect(apiRequest.ok()).toBeTruthy()
+        const response = await apiRequest.json()
+        const success = response.success
+        let checkSucces = (currentStatus !== newStatus) ? true : false
```
[the remark this answers is not in the document]
Comment (YulianH) [reply pages/apiPages/moderator_page.ts:79]: done

File: tests/api_tests/api_moderator_test.spec.ts (hunk 3 above)
Comment (YulianH) [reply tests/api_tests/api_moderator_test.spec.ts:6]: done

File: tests/api_tests/api_moderator_test.spec.ts (hunk 4 above)
Comment (YulianH) [reply tests/api_tests/api_moderator_test.spec.ts:8]: done

File: pages/apiPages/moderator_page.ts (hunk 5 above)
Comment (YulianH) [reply pages/apiPages/moderator_page.ts:65]: done

File: tests/api_tests/api_moderator_test.spec.ts (hunk 7 above)
Comment (YulianH) [reply tests/api_tests/api_moderator_test.spec.ts:45]: done

File: tests/api_tests/api_moderator_test.spec.ts (hunk 8 above)
Comment (YulianH) [reply tests/api_tests/api_moderator_test.spec.ts:62]: done

File: pages/apiPages/moderator_page.ts [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+        const response = await apiRequest.json()
+        const id = response.profile._id
+        const rolesGroup = response.profile.rolesGroup
+        const newAdminToken = response.token
+        expect(rolesGroup).toEqual('admin')
+        console.log(`Admin with: ${id}, ${email} and "password" is Logged In`)
+        return { id, email, newAdminToken }
+    }
+
+
```
[the remark this answers is not in the document]
Comment (YulianH) [reply pages/apiPages/moderator_page.ts:47]: done

File: tests/api_tests/api_referal_test.spec.ts (hunk 1 above)
Comment (YulianH) [reply tests/api_tests/api_referal_test.spec.ts:31]: done

File: pages/apiPages/referal_page.ts (hunk 2 above)
Comment (YulianH) [reply pages/apiPages/referal_page.ts:46]: done

**Pull request merged**